Feature/barcode scanner

jdavisAR opened the pull request:
# Changelog
- Added support for scanning QR codes and Barcode. (Natively supports opening deep-links embedded in QR code).
- Added an example permission request flow using composables.

## Added
- app/src/main/java/com/atomicrobot/carbon/ui/compose/CarbonCompositionLocals.kt
- app/src/main/java/com/atomicrobot/carbon/ui/compose/ComposePreviewParameters.kt
- app/src/main/java/com/atomicrobot/carbon/ui/compose/RequestPermission.kt
- app/src/main/java/com/atomicrobot/carbon/ui/scanner/ScannerScreen.kt
- app/src/main/java/com/atomicrobot/carbon/ui/scanner/ScannerViewModel.kt
- app/src/main/java/com/atomicrobot/carbon/util/Common.kt
- app/src/test/qr-codes/qr-code balck_18.png
- app/src/test/qr-codes/qr-code blue_48.png
- app/src/test/qr-codes/qr-code green_28.png
- app/src/test/qr-codes/qr-code red_22.png
 
## Changed
- app/build.gradle
- app/src/main/AndroidManifest.xml
- app/src/main/java/com/atomicrobot/carbon/StartActivity.kt
- app/src/main/java/com/atomicrobot/carbon/data/DataModule.kt
- app/src/main/java/com/atomicrobot/carbon/navigation/AppScreens.kt
- app/src/main/java/com/atomicrobot/carbon/navigation/Drawer.kt
- app/src/main/java/com/atomicrobot/carbon/ui/compose/MainNavigation.kt
- app/src/main/res/values/strings.xml
- build.gradle

## Is there test coverage for your changes?
 No

## Screenshots

jdavisAR committed:
- Started implementing Barcode/QR code scanner screen.

jdavisAR committed:
- Started implementing camera selector state.

jdavisAR committed:
- Detects barcodes in the live camera feed.

jdavisAR committed:
- Improved barcode scanning.
- Added an overlay to the screen to highlight barcode.

jdavisAR committed:
- Fixed deep linking.
- Fixed app behavior when the back action is fired.
- Deep-links encoded into a barcode can be opened.

jdavisAR committed:
- Added test QR codes.

jdavisAR committed:
Merge branch 'reference' into feature/barcode-scanner

bradyjohnk reviewed (changes requested):
Minor mis-type on one of the image files.
I suggest not using spaces in image file names. Use "_" instead.
Everything else looks good. Will do some testing and re-review again after changes are pushed

jdavisAR committed:
- Removed unused preview parameter provider.
- Fixed file names for sample qr codes.

bradyjohnk commented on app/src/main/java/com/atomicrobot/carbon/ui/scanner/ScannerViewModel.kt:
```diff
@@ -0,0 +1,209 @@
+    }
+}
+
+data class BarcodeAnalysisState(
+    val sourceImageWidth: Int = 0,
+    val sourceImageHeight: Int = 0,
+    val isFlipped: Boolean = false,
+    val barcode: Barcode? = null
+) {
+    val hasBarcodes: Boolean = barcode != null
+}
```
Let's move data class and sealed class to the top of the file.  To be consistent in the project (similar to MainViewModel).
If it's not a part of ScannerViewModel then it should be it's own file somewhere in /data/

jdavisAR replied:
I've moved the data class to top of the file to match the style in MainViewModel.

bradyjohnk reviewed (changes requested):
Another question. Does this only do QR codes or also barcodes?
If it does barcodes as well I think we should also provide an image example to test against.

bradyjohnk commented on app/src/main/java/com/atomicrobot/carbon/ui/scanner/ScannerViewModel.kt:
```diff
@@ -0,0 +1,209 @@
+    private val isFrontFacing
+        get() = _cameraSelectorState.value.lensFacing == CameraSelector.LENS_FACING_FRONT
+
+    private val isBackFacing
+        get() = isFrontFacing
+
```
I think this should be using the "!isFrontFacing"? Or should there be a hard check against LENS_FACING_BACK? in case there are more than 2 options on a future phone?

jdavisAR replied:
This was a typo but since the property wasn't needed I've removed it.

bradyjohnk commented on app/src/main/java/com/atomicrobot/carbon/ui/scanner/ScannerViewModel.kt:
```diff
@@ -0,0 +1,209 @@
+    private val isBackFacing
+        get() = isFrontFacing
+
+    private val isImageFlipped
+        get() = isFrontFacing
+
```
I think this is okay for our app. But maybe we don't need this variable? We can just check if it's front facing and we know that it is flipped in our app?
(thinking that if the user was able to flip the image for each camera then the check would need to be different. I know we don't do that for this app,  but this logic/naming doesn't seem to be helpful? Or maybe we just leave a comment next to it explaining)

jdavisAR replied:
Removed

jdavisAR committed:
- Removed unnecessary properties from ScannerViewModel.kt
- Scanner overlay no longer draws barcode text to screen.
- Renamed the folder where test barcodes & QR-codes are saved.

jdavisAR commented:
> Another question. Does this only do QR codes or also barcodes? If it does barcodes as well I think we should also provide an image example to test against.

I've added a sample barcode that encodes the following text "AR 5155".
